### PR TITLE
Atmos Grenades (aka: how to kill people with atmos)

### PR DESCRIPTION
**Pull request by Kyep**
Adds atmos grenades, a boxed set of grenades traitors can get via an uplink. 

Each set includes one each of the following:
- A plasma fire grenade, which creates a plasma fire
- A knockout grenade, which spreads N2O throughout an area

They use cluster grenade mechanics to spread more quickly than normal gasses do.

These grenades make it actually viable to kill people with atmos, to deny usage of an area using atmos, to knock people out using atmos, etc.

What purpose does this serve?
- Lets traitors kill people using atmos
- Lets traitors deny (or make difficult) usage of areas (e.g: blocking off a path behind them, trapping someone in an area, etc)
- Gives atmos techs the possibility that they might actually be needed to fix something

:cl: Kyep
add: Added atmos grenade kit to traitor uplink.
/:cl:

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Prepare to get shanked by @tigercat2000 

Just sayin'.

**Comment by Kyep** [comment]
@Fox-McCloud So, life as usual?

**Comment by IcyV** [comment]
These sound really cool. 6TC almost feels a little underwhelming, but it's hard to say. It also seems like the suffocation/vacuum grenades will end up serving essentially the same purpose in the end though. 

How exactly will they end up working though? Will it be affected by walls/doors/etc? Will the plasma fire one instantly place the plasma and light it all at once, or light from the detonation point? Is there a specific amount of gas each grenade will produce per tile? 

I like the sound of them overall but some numbers and exact details in your perspective of them would be cool.

**Comment by Kyep** [comment]
@IcyV to answer your questions:
- 6TC is because they're probably better than the syndicate minibomb, which is also 6 TC. (though, one could argue, as I would, the minibomb is overpriced)
- Suffocation is meant to be a stealth kill, in that it provides little warning, but is easily countered with internals if you're watching for it. Vacuum is intended to be 'screw this area'. Hardsuit required. 
- Yes, the code takes walls/doors/etc into account. The initial 20 tile detonation area cannot cross any wall, closed door, glass barrier, or anything else that blocks atmos.
- Yes, the plasma fire ignites the entire 20 tile initial area instantly (or as close to instantly as BYOND processing works) and spreads from there at normal speed.
- Yes, there is a specific amount each gas grenade adds to each tile - but it isn't a fixed total divided by the number of tiles - it is a static number added to each tile over its up-to-20 initial effect area. Meaning if you detonate it in a 6-tile room it adds X to each tile. If you detonate it in a closed 20 tile room, it still adds X to each tile. If you detonate it in a 40 tile room, it adds X to 20 tiles, then it spreads from there.

The exact numbers are in the 'files changed' section, but I will freely admit they're not final. I'd really prefer someone more experienced than I with atmos to give them a look over.

**Comment by IcyV** [comment]
Yeah, I didn't look too hard into the files itself mostly for your interpretation of them all at the moment. I don't really see anything wrong with any of this though. I think suffocation and vacuum *could* be fairly weak compared to the others, but that would be up to how someone decides to use it. Looking forward to seeing how it develops.

**Comment by Kyep** [comment]
Right now, the various grenades are mostly there as a proof-of-concept, so people can give feedback.
If people think that some of them would be useless, or that the whole set is too OP, or whatnot, I can take some out.

**Comment by CrAzYPiLoT-SS13** [comment]
I really like it. Finally, some help as a techie. :+1:

File: code/game/objects/items/weapons/grenades/atmosgrenade.dm [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	gas_type = "knockout"
+
+/obj/item/weapon/grenade/gas/prime()
+	var/turf/simulated/origin = get_turf(src)
+	var/list/spread_turfs = list()
+	var/list/scan_turfs = list()
+
+	if(!origin)
+		// Error, unsimulated turf.
+		return
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]: This won't stop it from working on an unsimulated turf - you'll want an `istype` for that

Comment (Kyep) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]: Thanks Crazy. Changed.

**Comment by Fox-McCloud** [comment]
I'll like @tigercat2000 pick at this, but.

If I'm understanding this correct.....Instantaneously converting all turfs in an area to CO2 or NO2 is overpowered as hell, especially since CO2 is lethal, too.

**Comment by IcyV** [comment]
>If I'm understanding this correct.....Instantaneously converting all turfs in an area to CO2 or NO2 is overpowered as hell, especially since CO2 is lethal, too.

It's fairly strong if prepared correctly, probably not just on the fly though. Given certain species probably won't be affected in the first place and that everyone will likely have access to immediate internals the lethal side of it probably minimal. The N02 can probably be fairly strong if people don't leave in time, but it's immediately countered by anyone that runs in with internals already on or taking a person outside of the room. 

Full instant conversion prooobably is on the strong side for any gas though. Could be more balanced if it just semi-quickly converts outwards from the center point.

File: code/game/objects/items/weapons/grenades/atmosgrenade.dm [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	scan_turfs += origin
+
+	while(spread_turfs.len < max_spread_turfs && scan_turfs.len)
+		if(gas_debug)
+			visible_message("[src] loops: [spread_turfs.len] turfs, [scan_turfs.len] to scan!")
+		var/turf/scan_from = pick(scan_turfs)
+		var/list/adj = 	scan_from.GetAtmosAdjacentTurfs(1)
+		for(var/turf/simulated/C in adj)
+			if(!(C in scan_turfs) && C.CanAtmosPass(C))
```
Comment (monster860) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]:
```
if(!(C in scan_turfs) && !(C in spread_turfs) && C.CanAtmosPass(C))
```

File: code/game/objects/items/weapons/grenades/atmosgrenade.dm [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		scan_turfs -= scan_from
+
+	if(!spread_turfs.len)
+		visible_message("<span class='notice'>[src] detonates harmlessly in the airless space.</span>")
+		qdel(src)
+	else
+		if(gas_debug)
+			visible_message("[src] affects [spread_turfs.len] turfs, with [scan_turfs.len] left to scan!")
+		for(var/turf/simulated/target_turf in spread_turfs)
+			switch(gas_type)
```
Comment (monster860) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]: SNOWFLEKED CODE!!!
===

Instead, create a proc which is overridden by the different grenade subtypes.

File: code/datums/uplink_item.dm [hunk 4]
```diff
@@ -452,6 +452,13 @@ var/list/uplink_items = list()
 	gamemodes = list(/datum/game_mode/nuclear)
 	surplus = 0
 
+/datum/uplink_item/dangerous/atmosgrenades
+	name = "Atmos Grenades"
+	desc = "A box of four (4) grenades that wreak havoc with the atmosphere of the target area, causing fire, suffocation, vacuum and/or crushing pressure. Use extreme caution! Unlike normal gasses, these are explosively propelled and instantly engulf a large area!"
```
Comment (monster860) [inline code/datums/uplink_item.dm]: :eyes:

Description says 4 grenades, but box actually starts with 5.

**Comment by Belario** [comment]
:+1: Because atmos need a reason to exist.

I'm not so sure about the disabling air alarms though, instantly effecting all those turfs all at once could take out dozens if it's aimed properly, should maybe be something more shockingly obvious due to it's great potency.

File: code/game/objects/items/weapons/grenades/atmosgrenade.dm [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+					target_turf.atmos_spawn_air(SPAWN_20C | SPAWN_N2O, 100) // knocks out quickly
+				if("vacuum")
+					target_turf.air.nitrogen = 0
+					target_turf.air.oxygen = 0
+					target_turf.air.nitrogen = 0
+					air_master.add_to_active(target_turf, 0)
+			var/obj/machinery/alarm/air_alarm = locate(/obj/machinery/alarm/) in target_turf
+			if(istype(air_alarm))
+				// break the alarm off the wall
+				air_alarm.visible_message("[air_alarm] breaks from the rapid change in pressure!")
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]: Span class.

**Comment by tigercat2000** [comment]
I'm 110% against this. :-1:

Your scanning method is incredibly flawed. It randomly picks a turf out of every one scanned so far and arbitrarily adds all adjacent turfs to pick from again. CanAtmosPass() shouldn't be called so lightly either.

That's not to mention how ridiculously powerful it is to release CO2 or N2O in a room instantly.

**Comment by Allfd** [comment]
I think maybe it should spread out not instantly, sort of like
 

Perhaps the N2O one could be used by security for riots.  Would probably be very ineffective but would look cool.

**Comment by theColdflame** [comment]
The idea's good but this seems like an 'i click u ded' tool.

**Comment by monster860** [comment]
So? EMP's are also an 'i click u ded' tool if you're an IPC.

**Comment by tigercat2000** [comment]
... so what
that doesn't mean every other race gets a "fuck you you're dead" downside @monster860

**Comment by SpaceTimeNow** [comment]
So we implement this weak excuse of an atmos system and then have to create grenades that circumvent it because our atmos now sucks. Really? That's just brilliant mate! 

#LindaStillSucks

**Comment by Da-Dman234** [comment]
I fail to see how this is any more overpowered than a regular grenade you throw, which causes dismemberment and most likely instant death. Atmos techs need toys to play with too, and none of these grenades, bar the N20 one, seem especially strong. Not to  mention the fact that once one of these are thrown all you need to counter the majority of them are use the supplies you find in your emergency equipment box.

**Comment by Kyep** [comment]
@tigercat2000 
MAJOR CHANGES:
1) Only Plasma Fire and N2O Knockout grenades exist now.
2) Both types now use the clusterbuster mechanic, rather than new ZAS-like spread code, to distribute their gasses. This means their spread is slower, but they can potentially cover more area. It also means they detonate slower - you have more time to get out of the way after they're thrown.
3) You only get one of each type of grenade in a kit.
4) Massive refactor and elimination of code.

**Comment by GelatinousGlob** [comment]
Would it be possible to make an atmos grenade that can be filled with an arbitrary gas mix that an atmosian makes?

**Comment by tigercat2000** [comment]
@Kyep All is good, then.

File: code/game/objects/items/weapons/grenades/atmosgrenade.dm [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	if(istype(target_turf))
+		target_turf.atmos_spawn_air(SPAWN_HEAT | SPAWN_TOXINS, 100)
+		target_turf.air_update_turf()
+	qdel(src)
+
+/obj/item/weapon/grenade/gas/knockout
+	name = "Knockout Grenade"
+	desc = "A grenade that completely removes all air and heat from its detonation area."
+
+/obj/item/weapon/grenade/gas/knockout/prime()
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/grenades/atmosgrenade.dm]: the copy-paste.

Make this more generic and modular.

**Comment by Kyep** [comment]
@GelatinousGlob Maybe, but:
1) I don't want to.
2) I would have balance concerns.
3) I am trying to keep this simple and predictable.
4) Such grenades would really only be useful for atmos techs.

@Fox-McCloud 
Refactored

**Comment by Fox-McCloud** [comment]
That's better.

File: code/game/objects/items/weapons/storage/uplink_kits.dm [hunk 7]
```diff
@@ -186,4 +186,13 @@
...
+	new /obj/item/weapon/gun/projectile/revolver/doublebarrel/improvised/cane(src)
+
+
+/obj/item/weapon/storage/box/syndie_kit/atmosgasgrenades
+	name = "Atmos Grenades"
+
+/obj/item/weapon/storage/box/syndie_kit/atmosgasgrenades/New()
+		..()
+		new /obj/item/weapon/grenade/clusterbuster/plasma(src)
+		new /obj/item/weapon/grenade/clusterbuster/n2o(src)
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/storage/uplink_kits.dm]: dat indentation

Comment (TheDZD) [inline code/game/objects/items/weapons/storage/uplink_kits.dm]: Yeah, should only be indented once, not twice.

**Comment by Necaladun** [comment]
...personally I've found N2O flamethrowers a pretty good way to fill a room quickly. CO2 ones, not so good...working on the mix.

If it converts all air in the room - does it keep the same other qualities like air temp?

**Comment by Kyep** [comment]
@Necaladun the conversion grenade was removed in the last major update. Only fire and N2O grenades exist now.

**Comment by GelatinousGlob** [comment]
I still think custom grenades would be good. If you allow these grenades then can traitor atmos spew plasma cans and shit without hijack?

**Pull request merged**